liferay-echo/liferay-portal
# [TS] LPS-194075 Menu Display widget yields misleading configuration when placed on a page template

Ticket: [LPS-194075](https://liferay.atlassian.net/browse/LPS-194075)

Motivation
=======
Currently, the Menu Display widget's default configuration is misleading when placed on a Page Template. The default configuration will be either "Private Pages Hierarchy" if there are any Private Pages on the Site, or it will be "Primary Navigation Menu" if there aren't any Private Pages on the Site. It will never be "Public Pages Hierarchy" because Page Templates are considered as Private Pages.

If the configuration of the Menu Display widget is never modified, then it will be considered to be in the "default" configuration. This means that when a Public Page inherits the Page Template, the Menu Display will render the "Public Pages Hierarchy", even though on the Page Template it was rendering either the "Primary Navigation Menu" or "Private Pages Hierarchy" instead. This can pose a nasty surprise to the user if they were expecting it to render the "Primary Navigation Menu" as the Page Template seemed to show.

Furthermore, if there are Private Pages on the Site, the user is able to explicitly configure the Menu Display on the Page Template for "Private Pages Hierarchy" by changing the configuration away from "Private Pages Hierarchy" and then changing it back to "Private Pages Hierarchy". This means that if a Public Page inherits the Page Template, it will display the "Private Pages Hierarchy", which is something that should not be allowed on Public Pages.

Proposed Solution
========
I changed the default configuration such that if a Menu Display widget is placed on a Page Template, the default configuration for it will be "Pages Hierarchy" (unless there are no Pages at all on the Site). It will display as "Pages Hierarchy" even if Private Pages are enabled. If Private Pages are enabled, an alert will appear on the Page Template letting the user know that the Menu Display may appear different if the inherited page has the opposite `privateLayout` value to the menu currently being rendered (see screenshot below).

By default, the Menu Display widget on the Page Template will display the Public Pages Hierarchy. However, if there are no Public Pages (but there are Private Pages) on the Site, then it will display the Private Pages Hierarchy instead. If there are no pages at all, then the "Pages Hierarchy" option will not appear, and the "Primary Navigation Menu" will be the default option (similar to how it already works on non-PageTemplate pages).

Steps to Verify
========
1. Start up Liferay and log in as the admin user.
2. Navigate to Site Builder > Navigation Menus.
3. Add a new Navigation Menu as follows:
* Name it "Test Navigation Menu"
* Add a single Page element to it representing the "Home" page. 
* Click on "Home" page element's three-dot menu > View Info. Check the "Use Custom Name" box and rename it to "Test". Save.
* Navigate back to the Navigation Menus page and click on "Test Navigation Menu"'s three-dot menu > Mark as Primary Navigation.
4. Navigate to Design > Page Templates. Add a new Master Page as follows:
* Name it "Test Master Page"
* Add a new "Menu Display" widget (NOT the fragment, but the widget). Observe that the "Menu Display" widget shows "Test", since it is configured to display the Primary Navigation Menu by default.
* Click the Publish Master button.
5. Navigate back to the "Home" page and edit it as follows:
* Click the "Page Design Options" icon (it is the icon that looks like a paint roller).
* Set the Master to "Test Master Page”.

**Expected Result:** The Navigation Menu widget that appears on the page would be showing "Test", since that's what it was configured to show on the Master that is being inherited.
**Actual Result:** The Navigation Menu widget that appears on the page is showing "Home".

Screenshots (if applies)
========
Below is a screenshot of what the "Default Configuration" now looks like on a Page Template with Private Pages enabled: 
<img width="1920" alt="LPS-194075 Menu Display with Info Alert" src="https://github.com/liferay-echo/liferay-portal/assets/15066512/12d1f445-0214-4edb-b5b1-cfcb7d7aea06">

When a Page inherits this Page Template, the alert message does not display on the Page, and the Navigation Menu will be configured for either Public Pages Hierarchy or Private Pages Hierarchy depending on whether the page is public.
